xcp_d/utils/utils.py

```diff
@@ -463,6 +463,36 @@ def denoise_with_nilearn(
         sample_mask=sample_mask,
         t_r=TR,
     )
+    # Replace any high-motion volumes at the beginning or end of the run with the closest
+    # low-motion volume's data.
+    # From https://stackoverflow.com/a/48106843/2589328
+    outlier_idx = sorted(set(np.where(~sample_mask)[0]))
+    if outlier_idx:
+        gaps = [[s, e] for s, e in zip(outlier_idx, outlier_idx[1:]) if s + 1 < e]
+        edges = iter(outlier_idx[:1] + sum(gaps, []) + outlier_idx[-1:])
+        consecutive_outliers_idx = list(zip(edges, edges))
+        first_outliers = consecutive_outliers_idx[0]
+        last_outliers = consecutive_outliers_idx[-1]
+
+        # Replace outliers at beginning of run
+        if first_outliers[0] == 0:
+            LOGGER.warning(
+                f"Outlier volumes at beginning of run ({first_outliers[0]}-{first_outliers[1]}) "
+                "will be replaced with first non-outlier volume's values."
+            )
+            interpolated_unfiltered_bold[
+                : first_outliers[1] + 1, :
+            ] = interpolated_unfiltered_bold[first_outliers[1] + 1, :]
+
+        # Replace outliers at end of run
+        if last_outliers[1] == n_volumes - 1:
```

Comment:
The problem only arises if the very last or very first points are censored?

Reply:
Yeah, it seems like cubic spline interpolation works well as long as there's some data before and after the points being interpolated, so the problem comes from interpolating with point only on one side.